File: z/平衡二叉树.cc
class Solution {
 public:
  int max_depth(TreeNode* root) {
    if (root == NULL) {
      return 0;
    }
    return max(max_depth(root->left), max_depth(root->right)) + 1;
  }
  bool IsBalanced_Solution(TreeNode* pRoot) {
    if (pRoot == NULL) {
      return true;
    }
    if (abs(max_depth(pRoot->left) - max_depth(pRoot->right)) > 1) {
      return false;
    }
    bool left = IsBalanced_Solution(pRoot->left);
    bool right = IsBalanced_Solution(pRoot->right);
    return left && right;
  }
};

File: h/平衡二叉树.java
/*
输入一棵二叉树，判断该二叉树是否是平衡二叉树。
*/
public class Solution {
    public boolean IsBalanced_Solution(TreeNode root) {
        if (root == null){
            return true;
        }
        int left = iterTree(root.left);
        int right = iterTree(root.right);
        
        if (Math.abs(left - right) > 1) {
            return false;
        }
        return true;
    }
    
    private static int iterTree(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int left = iterTree(root.left);
        int right = iterTree(root.right);
        
        return left > right ? (left + 1) : (right + 1);
    }
}
